lib/readability/published_at_finder.ex

```diff
@@ -0,0 +1,88 @@
+defmodule Readability.PublishedAtFinder do
+  @moduledoc """
+  Extract the published at.
+  """
+
+  @type html_tree :: tuple | list
+
+  @strategies [:meta_tag, :time_element, :data_attribute, :text]
+
+  @doc """
+  Extract the published at.
+  """
+  @spec find(html_tree) :: %DateTime{} | %Date{} | nil
+  def find(html_tree) do
+    value =
+      Enum.find_value(@strategies, fn strategy ->
+        strategy(strategy, html_tree)
+      end)
+
+    if value do
+      parse(value)
+    end
+  end
+
+  defp strategy(:meta_tag, html_tree) do
+    selector = "meta[property='article:published_time'], meta[property='article:published']"
+
+    html_tree
+    |> Floki.find(selector)
+    |> Enum.flat_map(&Floki.attribute(&1, "content"))
+    |> Enum.map(&String.trim/1)
+    |> List.first()
+  end
+
+  defp strategy(:time_element, html_tree) do
+    html_tree
+    |> Floki.find("time")
+    |> Enum.flat_map(&Floki.attribute(&1, "datetime"))
+    |> Enum.map(&String.trim/1)
+    |> List.first()
+  end
+
+  defp strategy(:data_attribute, html_tree) do
+    html_tree
+    |> Floki.find("[data-datetime]")
+    |> Enum.flat_map(&Floki.attribute(&1, "data-datetime"))
+    |> Enum.map(&String.trim/1)
+    |> List.first()
+  end
+
+  defp strategy(:text, html_tree) do
+    html_tree
+    |> Floki.find("div, p, span")
+    |> Enum.map(&Floki.text/1)
+    |> Enum.reject(&(String.length(&1) < 4))
+    |> Enum.map(&parse(:string, &1))
+    |> Enum.filter(& &1)
+    |> Enum.map(&to_string/1)
+    |> List.first()
+  end
+
+  defp parse(value) do
+    Enum.find_value([:datetime, :date, :string], fn parser ->
+      parse(parser, value)
+    end)
```

Comment:
It took me a moment to understand how find_value exactly works with that approach.

Why not

```
parse(:datetime, value) || parse(:date, value) || parse(:string, value)
```

or maybe even (just braindumping, not sure about that :D) 

```
with {:error, _} <- DateTime.from_iso8601(value),
     {:error, _} <- Date.from_iso8601(value), 
     {:error, _} <- DateTimeParser.parse(value) do
     nil
else 
     {:ok, datetime, _} -> datetime
     {:ok, datetime} -> datetime
     _ -> nil
 end       
```

Probably separate functions is clearer and you can use only one of them 😅

Reply:
Yeah you're right, I've changed it to:

```elixir
parse(:datetime, value) || parse(:date, value) || parse(:string, value)
```

I'm not a fan of `with..else`